Add Basics tutorial page

shogowada opened the pull request:
I believe majority of people coming to tutorials are more interested in quick examples than detailed explanations.

The current first real tutorial is Unified Types. While it's helpful to know that information, it doesn't give you much insight into how Scala code would look like. I think it can lose interest of readers.

Also, while many pages are using features like `object`, `App`, `var`, `val`, etc, what they are isn't necessarily clearly communicated before it's introduced. I think it can throw out readers.

I created a Basics page to address those concerns. The page is supposed to provide just the most important information out of the most important topics.

Personally, for this kind of tutorial, I think it makes more sense to use REPL because it will be easier for writers to make sure what they provide is completely reproducible on reader's machine, and it will be more interesting to readers because it will be easier to interact and play with the code. While I didn't do it because it will be a bigger change if you consider consistency, I think it's worth considering.

Another option is to give them link to the [ScalaFiddle](https://scalafiddle.io) so that we can achieve the same goal without readers installing anything on their machine.

I tried to test it locally, but I am having trouble setting up the build environment (`bundle install` fails because `gem install json -v '1.8.3'` fails, tried multiple solutions found online, and eventually gave up), so I haven't tested it yet, but it looks fine on GitHub.

MasseGuillaume commented on tutorials/tour/basics.md:
````diff
@@ -0,0 +1,207 @@
+}
+```
+
+Because every runnable application has a main method, Scala also provides a more convenient way of defining your main method by extending `App` trait.
````
This trait is deprecated. (App -> DelayedInit)

http://www.scala-lang.org/api/current/scala/DelayedInit.html

(Since version 2.11.0) DelayedInit semantics can be surprising. Support for App will continue. See the release notes for more details: https://github.com/scala/scala/releases/tag/v2.11.0-RC1

jvican replied:
This trait is not deprecated. It will be maintained both in future Scala and Dotty versions.

MasseGuillaume replied:
You don't want to start an application inside a constructor.

tpolecat replied:
Agree, `App` is problematic due to `DelayedInit`. I suggest not mentioning it.

MasseGuillaume commented:
That's an awesome start!

jvican commented on tutorials/tour/basics.md:
````diff
@@ -0,0 +1,207 @@
+x += 1
+```
+
+Often, you want your variables to be immutable. You can use `val` keyword in that case.
````
What about "Mutable variables are not idiomatic Scala and therefore discouraged. Most of the times, you can use `val` definitions."?

a1russell replied:
@jvican That wouldn't tell the newcomer that `val` defines an immutable variable. I also don't think we should conflate the goals of a style guide (to explain "idiomatic" Scala) with what should simply be the introduction to the language features.

heathermiller replied:
@a1russell, I don't necessarily agree with this sentiment. Because you can write code that is essentially Java with a `.scala` extension and pass it through the Scala compiler, I think it's extremely important to focus on the basics of what Scala really is in this article. When it comes to immutable vs mutable references (`val` vs `var), Scala is all about `val`s. It's rare to have to use a `var` anywhere in a normal Scala application. Therefore, I think introducing `val`s should probably come first, and `var`s should probably be introduced as that thing from Java that you can use if you absolutely have to, but is generally avoided.

a1russell replied:
@heathermiller I agree that `val` should probably be introduced first. My issue was with @jvican's proposed wording is all.

jvican replied:
It's not only that it should be introduced first, it should be emphasized as the de-facto alternative when declaring variables, @heathermiller conveys very well why. @a1russell

tpolecat replied:
`var` is an advanced feature that should not be mentioned at all in a beginner document, or should be mentioned only in passing.

a1russell replied:
I was thinking of something more along the lines of:

> You can define values with the `val` keyword.
>
> ```
> val x = 1 + 1
> ```
>
> Values cannot be re-assigned and are immutable in that sense. If you do require a mutable variable, however, you can use the `var` keyword. However, `val` should be used by default since `var` generally indicates that it *will* be mutated down the line.
>
> ```
> var y = 1 + 2
> y += 4
> ```

Feel free to further clean up the wording, but really all I meant was that I think we should avoid hand-wavy words like "idiomatic", and when we introduce something, we should briefly describe what it is.

ivanopagano replied:
I'm not sure that _immutable_ should be used at all. I (maybe incorrectly) expect that newcomers will be later confused by the whole immutable reference vs. immutable object argument, since it's not immediately intuitive that immutability at the variable definition level doesn't guarantee immutability of _what_ is referenced by that variable name.

@a1russell your definition seems good to me in spirit, but I'd rephrase the second part as

> `val`s cannot be re-assigned and they will always reference the same value. If you need to re-assign a variable, you can use the `var` keyword. It is considered good practice to reduce the use of `var`s to the bare minimum.
  ```
  var y = 1 + 2
  y += 4
  //y now is equal to 7
  ```

jvican commented on tutorials/tour/basics.md:
````diff
@@ -0,0 +1,207 @@
+(name: String) => println("Hello, " + name + "!")
+```
+
+You can also assign functions to variables.
````
Can you explain the different syntactic parts of a function? Signature and body? This is good :smile:, but I think those parts should be brought to the attention of the reader.

shogowada replied:
Sure. Maybe my assumption of "most readers are already familiar with anonymous function" is wrong. Even if it was right, I agree it's better be explained. Let me think of a simpler way of introducing functions.

jvican commented on tutorials/tour/basics.md:
```diff
@@ -0,0 +1,207 @@
+    y = y + dy
+  }
+
+  override def toString: String = "(" + x + ", " + y + ")"
```
I would not write this in the class example, since case classes are already doing this for you. Do you think it's a good idea to remove it and add the following to the case class section?
> What if we want to `println` a case class? The Scala compiler automatically writes it for you, so it's not necessary that you override `toString`.

tpolecat replied:
I would start with `case class` … non-`case` classes are relatively rare in Scala code. `new` and reference equality are advanced feature that should be de-emphasized.

tpolecat replied:
You are using `def` and `override` which you have not yet introduced, as well as the constructor syntax which is unusual when compared with other OO languages. I would explain these.

shogowada replied:
@jvican 

> I would not write this in the class example, since case classes are already doing this for you.

I agree. Let me change the example to something more suitable for classes.

@tpolecat 

> I would start with case class … non-case classes are relatively rare in Scala code.

We need to make this tutorial easy to understand for non-Scala developers. My concern is that introducing case classes first confuses them because case class is very specific to Scala.

> You are using def and override which you have not yet introduced, as well as the constructor syntax

Good point. I will add section for "Methods" to cover the `def` keyword and avoid using `override` here at all. I'll also add some explanation around constructor.

jvican commented on tutorials/tour/basics.md:
```diff
@@ -0,0 +1,207 @@
+
+## Case Classes
+
+By default, classes are compared by reference. If you want to compare them by values, you need to override `equals` and `hashCode` methods.
```
I would enumerate more clearly the benefits of a case class over a class, either here or at the end of this section. I think it's good to show them need by need, but they should be summarized somewhere :smile:.

tpolecat replied:
I would leave the `equality/hashCode` business until a chapter on classes. It's introducing a lot of new and problematic ideas that probably don't belong on page 1.

shogowada replied:
@jvican 

> I would enumerate more clearly the benefits of a case class over a class

While I agree there is a lot more to case class, I am also concerned mentioning things like pattern matching or even `copy` method (that I'm thinking to remove) can intimidate readers.

It is not my goal to give readers a complete picture of Scala. My goal here is to nudge readers into writing their first Scala code because I am convinced that once they start writing it, majority of them will have a positive experience.

Scala is a relatively complex language with many features. One common dilemma I see is that its own feature richness is intimidating newcomers, while the language is actually designed to make your code as simple as possible.

In the end of this section, I have the following sentence.

"There are many other features you get out-of-the-box by using case classes. We will cover them in depth [later](http://docs.scala-lang.org/tutorials/tour/case-classes)."

I might change the wording so that it is clearer that there is a lot more to case classes, but do you think having a link to the complete [Case Classes page](http://docs.scala-lang.org/tutorials/tour/case-classes) is good enough?

@tpolecat 

Agreed. Let me remove them.

jvican commented on tutorials/tour/basics.md:
````diff
@@ -0,0 +1,207 @@
+println(point == yetAnotherPoint) // false
+```
+
+Case classes are immutable by default, but it also provides `copy` method so that you can easily create another instance of the class while reusing the values from exiting instances.
````
"Case classes are immutable by default. They provide a `copy` method that allows you to modify any of the values defined in a case class."

shogowada replied:
It sounds better, but I think "modify any of the values defined in a case class" also sounds like it will mutate values, which is not what's happening. I cannot think of a simple way of wording what's actually happening, though. I started to think maybe `copy` is not worth mentioning in this page.

jvican commented on tutorials/tour/basics.md:
````diff
@@ -0,0 +1,207 @@
+println(point == yetAnotherPoint) // false
+```
+
+Because of case classes, you almost never see classes overriding `equals` and `hashCode` methods in Scala.
````
Case classes generate their own `equals` and `hashCode` methods so that you don't need to define them yourselves. If you want custom implementations, you can always override them.

heathermiller commented:
This is great @shogowada!! Would you be up for addressing some of the reviewer feedback?

heathermiller commented:
PS, I wish something like this existed when I was trying to learn Scala years ago. This is perfect for someone who already knows another language and who just wants to quickly be introduced to the "main Scala things one's gotta know about". This is great because it's so succinct and so concise; in literally 10 minutes, a Java developer could probably write a cute little Scala application. 

I love this ❤️

tpolecat commented on tutorials/tour/basics.md:
```diff
@@ -0,0 +1,207 @@
+
+## Singleton Objects
+
+You can define singleton objects with `object` keyword.
```
We should be careful about "singleton" because it alludes to the Go4 pattern which isn't quite right. I would say "`object` defines a module of definitions" or something like that.

Introducing global mutable state in an `object` is not a good example. Maybe just declare some values and methods?

shogowada replied:
> We should be careful about "singleton" because it alludes to the Go4 pattern which isn't quite right.

Can you elaborate? Do you mean it is not a singleton because the instance is not retrieved via method? I think "singleton" is more friendly than "a module of definitions" to most readers, and it's not wrong either. Do you agree?

> Introducing global mutable state in an object is not a good example.

I agree, but I couldn't think of better way of showing this object actually is a singleton without having states. Ideas?

a1russell replied:
"Module of definitions" sounds more like a trait to me, especially when reading it from the perspective of a Ruby developer.

tpolecat replied:
An `object` is a module. A namespace. Like a package, but unlike a package it's a value. When you say "singleton", as a Java programmer I think "singleton of what?" ... is `object Foo` an instance of `class Foo`? I realize "singleton" is the common usage but to me it's misleading.

shogowada replied:
I see. Let me name the section "Objects" instead of "Singleton Objects", then explain what it really is in details in the section.

tpolecat commented on tutorials/tour/basics.md:
```diff
@@ -0,0 +1,207 @@
+  var x: Int
+  var y: Int
+  
+  override def toString: String = "(" + x + ", " + y + ")"
```
Can we provide an implementation that's not an override, since we haven't talked about that yet? How about `def distanceFromOrigin: Double = math.hypot(x, y)`?

tpolecat commented on tutorials/tour/basics.md:
````diff
@@ -0,0 +1,207 @@
+You can extend traits with `extends` keyword.
+
+```
+class MyPoint(var x: Int, var y: Int) extends Point {
````
I recommend a case class here and `offset` that returns a new instance.

Sciss commented on tutorials/tour/basics.md:
````diff
@@ -0,0 +1,207 @@
+
+```
+object Main {
+  def main(args: Array[String]): Unit = {
````
How about removing the unnecessary braces? That also shows why regular syntax without procedure variant is useful:

```scala
  def main(args: Array[String]): Unit =
    println("Hello, world!")
```

a1russell replied:
To play Devil's advocate, much like the [1TBS](https://en.wikipedia.org/wiki/Indent_style#Variant:_1TBS_.28OTBS.29) vs [traditional K&R](https://en.wikipedia.org/wiki/Indent_style#K.26R_style) religious war, the argument could be made that consistency is preferable, and that braces should stay no matter whether the method is one or multiple lines.

I realize this is against the mainstream Scala community consensus and the [Scala style guide](http://docs.scala-lang.org/style/declarations#methods), however.

tpolecat replied:
Yes, +1. Remove unnecessary braces.

stephenjudkins commented:
I think we should move `var`, non-case classes, custom `toString`, and `override` out of the intro sections of this. None of these are really idiomatic Scala at this level. I'm OK with having them somewhere in the early sections, with a brief caveat that good Scala style is to avoid them in favor of immutable values and data types. I just don't think they should be presented so early, especially not before things like `val` and `case class`.

shogowada commented:
Thank you all for the feedbacks! It's really helpful. I'll revise the documentation based on the feedbacks.

I realized that I might not necessarily agree with what the "Basics" should be with everybody in the thread. So please let me explain where I'm coming from.

What I want to achieve here is to grow the size of Scala community by welcoming developers from diverse backgrounds and beliefs. Although the diversity might introduce some conflicts here and there, I believe those conflicts ultimately benefit the community as a whole by letting them think and innovate. After all, if two people are having the same thought, one of them is redundant.

I believe Scala has a great potential of becoming such a community because, IMO, the language itself is not so opinionated towards one paradigm or another; it supports both OOP and FP, both mutable and immutable objects/collections, etc without developers fighting with the language.

To welcome developers from various background, one thing I want to make sure in this Basics tutorial is not to intimidate them. To achieve that, I think we need to sacrifice some of the "Scala-ism" especially in this Basics page. In other words, it's not my intention to make this page "Scala FP Basics", "Scala OOP Basics", or even "Scala Best Practices". My intention is rather to make this page just "Scala Basics".

For example, there are many popular languages where `var` equivalent is default instead of `val` equivalent. While I agree that we should introduce `val` first (and I'll fix that!), I think not mentioning `var` at all confuses many readers. Worst case, it might mislead them Scala is too limited and opinionated for them, and I don't want that to happen. I'd like to introduce non-case classes in this page too for the same reason.

Given that, I hope you could understand where I am coming from and why I am trying to be make this page like how it is.

But I agree this page is currently introducing some unnecessarily things such as `override` or `toString`. I'll try to simplify them.

Thanks again for all the feedbacks!

shogowada commented on tutorials/tour/basics.md:
````diff
@@ -0,0 +1,207 @@
+You can define functions like below.
+
+```
+(name: String) => println("Hello, " + name + "!")
````
Good point. Let me address those concerns.

shogowada commented:
I think I've addressed most of the concerns. I'd appreciate if you could review again.

I also added ["Running Scala on Browser" section](https://github.com/scala/scala.github.com/pull/671/files#diff-c2f160f90dcc962112c59cf07bb40f00R15) at the top so that it will be super easy for readers to start writing their first Scala code.

Sciss commented:
On 03/02/17 03:42, Adam Russell wrote:
> To play Devil's advocate, much like the 1TBS
> <https://en.wikipedia.org/wiki/Indent_style#Variant:_1TBS_.28OTBS.29> vs
> traditional K&R <https://en.wikipedia.org/wiki/Indent_style#K.26R_style>
> religious war, the argument could be made that consistency is
> preferable, and that braces should stay no matter whether the method is
> one or multiple lines.
> 
> I realize this is against the mainstream Scala community consensus and
> the Scala style guide
> <http://docs.scala-lang.org/style/declarations#methods>, however.
> 

Well, the style guide is some authority ;)
I think Scala is about minimising ceremony, for me that clearly includes
dropping unnecessary braces and other visual clutter...

DarkDimius commented on tutorials/tour/basics.md:
```diff
@@ -0,0 +1,261 @@
+
+## Variables
+
+You can define variables with the `val` keyword.
```
Would something along the lines of
 "you can give a name to a computed expression using a keyword `val`"
be better?

tpolecat replied:
The `val` keyword binds an identifier to a computed value.

DarkDimius replied:
@tpolecat, that's precisely the version I started with in my mind, but then I changed to "gives name" because I think that binding is a more formal\scientific term, while "give name" would be more accessible.

shogowada replied:
Oh, I just didn't know the official term:) Thanks for pointing this out.

Maybe something like "You can define values (a.k.a. immutable variables) with the `val` keyword." will be correct and understandable? I see a lot of website referring `val` as immutable variables.

heathermiller replied:
I'd side with @DarkDimius and @shogowada on this one; it's just friendlier to leave the sciencespeak out here. While "blah keyword binds an identifier to a computed expression" is the terminology you'd use in a PL fundamentals course or in a paper, I think most developers would find this terminology alienating. 

So yeah, how about this average between @shogowada's and @DarkDimius's suggestions:

"You can give names a computed expression using the `val` keyword. We call these _values_. Importantly, values are immutable; that is, once a computed expression is named with the `val` keyword, it cannot be reassigned to refer to a different computed expression."

(PS, we shouldn't use "immutable variable" as it's not exactly correct terminology)

WDYT?

shogowada replied:
Sorry, I prematurely modified the line so now this comment is outdated. Let me start a new comment on the latest line.

shogowada replied:
Moved to https://github.com/scala/scala.github.com/pull/671/files#r99466445.

tpolecat commented on tutorials/tour/basics.md:
````diff
@@ -0,0 +1,261 @@
+
+```
+var y = 1 + 1
+y += 2 // This compiles because "y" is declared with "var" keyword.
````
It might be better here and in the case above to use simple assignment like `y = 42` rather than the compound operation `+=`. Or at least desugar it to `y = y + 2`.

tpolecat commented on tutorials/tour/basics.md:
````diff
@@ -0,0 +1,261 @@
+println(add(1, 2)) // 3
+```
+
+Or it can take no parameters.
````
To avoid confusion with no-arg `def`s we might want to say "an empty argument list".

shogowada replied:
Good point; it can be confusing.

I think methods taking multiple parameter groups is more advanced than function taking parameters. Maybe we should instead be explicit on methods section, explicitly stating difference between parameters and parameter groups. Then, readers can understand difference between "no parameters" vs. "no parameter groups" without confusing.

tpolecat commented on tutorials/tour/basics.md:
````diff
@@ -0,0 +1,261 @@
+println(getTheAnswer()) // 42
+```
+
+If your function body spans across multiple lines, you can wrap them with `{}`.
````
s/can/must/

shogowada commented on tutorials/tour/basics.md:
```diff
@@ -0,0 +1,262 @@
+
+## Values and Variables
+
+You can define values (a.k.a. immutable variables) with the `val` keyword.
```
Continued from https://github.com/scala/scala.github.com/pull/671#discussion_r99347365

It's tough:) So,

- we don't want to use the term "immutable variable" because it's technically incorrect.
- I personally think "computed expression" can be intimidating although it is correct.

Maybe we should have "Expressions" section before "Values and Variables", so that terms like "computed expression" or "result of the expression" won't be as intimidating. There, we can also state things like "almost everything in Scala is an expression" or "multiple expressions must be grouped by `{}`", and we will be able to use the term "expression" without intimidating readers for the rest of the tutorial. In the section, we can use `println` as a tool to show the result of each expression because we are not using REPL.

How do you guys think?

ivanopagano replied:
Why not simply?
> You can assign a name to the result of an expression, with the `val` keyword

tpolecat commented on tutorials/tour/basics.md:
```diff
@@ -84,36 +115,25 @@ val getTheAnswer = () => 42
 Methods look and behave very similar to functions, but there are a few key differences between them.
 
-Methods are defined like below with the `def` keyword followed by its name, parameter(s), return type, and body.
+Methods are defined like below with the `def` keyword followed by its name, a list of parameter groups, return type, and an expression.
```
It might be worth mentioning that (unlike functions) methods can also have *type* parameters.

shogowada replied:
I felt like type parameters should not be included in this Basics page because then we probably need to add it for classes, case classes, and traits section too, and we will have hard time explaining what it is especially for people coming from dynamically typed language backgrounds.

I think we probably should create a whole new page dedicated for type parameters. Right now the topic seems to be scattered across multiple pages.

shogowada commented:
Any other concerns? Let me know if I missed some, too.

a1russell commented on tutorials/tour/basics.md:
```diff
@@ -0,0 +1,294 @@
+
+Expressions are computable statements.
+
+You can output results of expressions using the `println`.
```
Probably don't need "the":
> You can output results of expressions using `println`.

a1russell commented on tutorials/tour/basics.md:
````diff
@@ -0,0 +1,294 @@
+You can output results of expressions using the `println`.
+
+```
+println(1) // 1
````
Do you think it's obvious enough that `//` indicates a comment?

heathermiller replied:
Syntax highlighting should make it pretty obvious. So I think we don't have to worry about this 🙂

shogowada replied:
Ah, I didn't even think of that. But I think you are right and we'd better explicitly state that because not all language use `//` for comments.

shogowada replied:
> Syntax highlighting should make it pretty obvious.

That's a good point too.

heathermiller commented:
🎉 Thank you so much for your contribution @shogowada!! This is really huge! We **really** appreciate it ❤️ 

Merging now. Any further feedback, we can deal with in a separate PR 🙂 

(I'll address @a1russell's comment about removing "the" in a follow-up commit)

a1russell commented on tutorials/tour/basics.md:
````diff
@@ -0,0 +1,294 @@
+// Point(1,2) and Point(2,2) are different.
+```
+
+There is a lot more to case classes that we'd like to introduce, and we are convinced you will fall in love with it! We will cover them in depth [later](case-classes.md).
````
Fall in love with *them*, I think.

heathermiller replied:
Thanks, I'll address this too 🙂
